### ISSUE TITLE
allOf is not working

### ISSUE THREAD

**Issue by ozguruysal** [Surnet/swagger-jsdoc#46]
Hi I'm testing with this example from [OpenAPI Specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#models-with-composition) but the result is an empty object in Swagger UI. 

Code:

```yaml
definitions:
  ErrorModel:
    type: object
    required:
    - message
    - code
    properties:
      message:
        type: string
      code:
        type: integer
        minimum: 100
        maximum: 600
  ExtendedErrorModel:
    allOf:
    - $ref: '#/definitions/ErrorModel'
    - type: object
      required:
      - rootCause
      properties:
        rootCause:
          type: string
```

Result:

![untitled-1](https://cloud.githubusercontent.com/assets/53624/21148953/8c5d9f92-c162-11e6-8bd3-18b972a2ae26.png)

**Comment by drGrove** [comment]
Hi @ozguruysal,

Can you provide a little more context on how you're referencing this?

I just attempted this with latest master and this seems to be working.

![161219000132](https://cloud.githubusercontent.com/assets/1441230/21305100/60884172-c57e-11e6-96d2-cb5f9b2b6f85.png)

```javascript
'use strict';
// This has the definition above and a function for creating said object
var error = require('./error'); 
/**
 * @swagger
 * /error:
 *   get:
 *     description: Returns an error
 *     responses:
 *       500:
 *         description: An error has occurred.
 *         schema:
 *           $ref: '#/definitions/ErrorWithCause'
 */
app.get('/error', function(req, res) {
  var err = new error.ErrorWithCause(
    500,
    'Something bad happend',
    'Route is bad'
  );
  res.status(500).send(err);
});
```

**Comment by ozguruysal** [comment]
Hi and apologies for not providing enough data. Here's a more detailed example of the issue.

```
// routes.js

/**
 * @swagger
 * definitions:
 *   Product:
 *     properties:
 *       product_id:
 *         type: string
 *         description: Product's Unique ID
 *       name:
 *         type: string
 *         description: Product name
 *   ProductDetail:
 *     allOf:
 *     - $ref: '#/definitions/Product'
 *       properties:
 *         sku:
 *           type: string
 *         category:
 *           type: string
 */

/**
 * @swagger
 * /products/{productId}:
 *   get:
 *     tags:
 *       - products
 *     summary: Gets a product
 *     parameters:
 *       - name: productId
 *         in: path
 *         type: string
 *         required: true
 *         description: Product ID to be fetched
 *     responses:
 *       200:
 *         description: A product object
 *         schema:
 *           type: array
 *           items:
 *             $ref: '#/definitions/ProductDetail'
 */
app.get('/products/:productId', function(req, res, next) {
    // ...
});
```
Result:

![allof-example](https://cloud.githubusercontent.com/assets/53624/21309269/588e2848-c5e6-11e6-9959-b28444c7caf6.png)

If you reference the `Product` definition in the documentation it works fine, however if you reference an extended definition which is `ProductDetail` in our example, it fails.

**Comment by drGrove** [comment]
Okay, so it appears to be an issue on the generation while running in the application.

A quick fix until this can be tracked down is to generate the JSON via our cli (which oddly is working fine). I will look deeper into this tonight.

**Comment by kalinchernev** [comment]
I tested the snippets in [the editor](http://editor.swagger.io/#/) and the results are similar: there is a warning about the `allOf` property. I don't know if that might be helpful feedback:

![allof](https://cloud.githubusercontent.com/assets/1923476/21348824/2773a3e0-c6af-11e6-83a9-99535df6d081.png)

This is what I had in the editor:

```yaml
---
swagger: '2.0'
info:
  version: 0.0.0
  title: Simple API
paths:
  /:
    get:
      responses:
        200:
          description: OK
  /products/{productId}:
    get:
      tags:
        - products
      summary: Gets a product
      parameters:
        - name: productId
          in: path
          type: string
          required: true
          description: Product ID to be fetched
      responses:
        200:
          description: A product object
          schema:
            type: array
            items:
              $ref: '#/definitions/ProductDetail'

definitions:
  Product:
    properties:
      product_id:
        type: string
        description: Product's Unique ID
      name:
        type: string
        description: Product name
  ProductDetail:
    allOf:
    - $ref: '#/definitions/Product'
      properties:
        sku:
          type: string
        category:
          type: string
```

**Comment by ozguruysal** [comment]
Adding `- type: object` after `- $ref: '#/definitions/Product'` gets rid of the warning in Swagger editor. Like:

```yaml
ProductDetail:
  allOf:
  - $ref: '#/definitions/Product'
  - type: object
```

But adding that doesn't make any change for Swagger UI. Still an ampty object is displayed.

**Comment by kalinchernev** [comment]
@drGrove do you think that might be related to [this topic](https://github.com/BigstickCarpet/swagger-parser/issues/51) part of the parser?

**Comment by drGrove** [comment]
@kalinchernev that could very likely be the case. I haven't had a chance to dig any deeper though.

**Comment by jogakdal** [comment]
@ozguruysal You may put the additional properties of derived definition(ProductDetail) under `type` field and it will work. like below.  
```javascript
/**
 * @swagger
 * definitions:
 *   Product:
 *     type: object
 *     properties:
 *       product_id:
 *         type: string
 *         description: Product's Unique ID
 *       name:
 *         type: string
 *         description: Product name
 *   ProductDetail:
 *     allOf:
 *     - $ref: '#/definitions/Product'
 *     - type: object
 *       properties:
 *         sku:
 *           type: string
 *         category:
 *           type: string
 */

```

**Comment by ozguruysal** [comment]
@jogakdal thank you. We've moved to GraphQL but if we need to use swagger again, I'll try your solution.

**Comment by ganeshmahajan1985** [comment]
Hi @ozguruysal 

According to my observation on small change required in the YAML you have written as below - 
```YAML
ProductDetail:
  allOf:
    - $ref: '#/definitions/Product'

```
Just indent the -$ref 
Hope this help. 
Thanks

**Comment by krishna-birla** [comment]
`anyOf`, `allOf` etc were added in open api 3.0, and are nor available in 2.0 version.
